bigquery/v2/query/query.go

```diff
@@ -0,0 +1,183 @@
+	cachedPageToken string
+	cachedRows      []*Row
+	cachedSchema    *schema
+}
+
+func newQueryJobFromQueryResponse(c *Client, res *bigquerypb.QueryResponse) (*Query, error) {
+	schema := newSchema(res.Schema)
+	q := &Query{
+		c:               c,
+		cachedSchema:    schema,
+		cachedPageToken: res.PageToken,
+	}
+	if res.TotalRows != nil {
+		q.cachedTotalRows = res.TotalRows.Value
+	}
+	if res.JobComplete != nil {
+		q.complete = res.JobComplete.Value
+	}
+	if res.Rows != nil {
+		var err error
+		q.cachedRows, err = fieldValueRowsToRowList(res.Rows, schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if res.JobReference != nil {
+		jobRef := res.JobReference
+		q.projectID = jobRef.ProjectId
+		q.jobID = jobRef.JobId
+		if jobRef.Location != nil {
+			q.location = jobRef.Location.GetValue()
+		}
+	}
+	return q, nil
+}
+
+func newQueryJobFromJob(c *Client, job *bigquerypb.Job) (*Query, error) {
+	return newQueryJobFromJobReference(c, nil, job.JobReference)
+}
+
+func newQueryJobFromJobReference(c *Client, schema *bigquerypb.TableSchema, jobRef *bigquerypb.JobReference) (*Query, error) {
+	res := &bigquerypb.QueryResponse{
+		Schema:       schema,
+		JobReference: jobRef,
+	}
+	return newQueryJobFromQueryResponse(c, res)
+}
+
+// state is one of a sequence of states that a Job progresses through as it is processed.
+type state = string
+
+const (
+	// Pending is a state that describes that the job is pending.
+	Pending state = "PENDING"
+	// Running is a state that describes that the job is running.
+	Running state = "RUNNING"
+	// Done is a state that describes that the job is done.
+	Done state = "DONE"
+)
+
+func (q *Query) checkStatus(ctx context.Context) error {
+	res, err := q.c.c.GetQueryResults(ctx, &bigquerypb.GetQueryResultsRequest{
+		ProjectId:  q.projectID,
+		JobId:      q.jobID,
+		Location:   q.location,
+		MaxResults: wrapperspb.UInt32(0),
+		FormatOptions: &bigquerypb.DataFormatOptions{
+			UseInt64Timestamp: true,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = q.consumeQueryResponse(res)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (q *Query) consumeQueryResponse(res *bigquerypb.GetQueryResultsResponse) error {
+	q.cachedPageToken = res.PageToken
+	schema := newSchema(res.Schema)
+	var err error
+	q.cachedRows, err = fieldValueRowsToRowList(res.Rows, schema)
+	if err != nil {
+		return err
+	}
+	q.cachedSchema = schema
+	q.cachedTotalRows = res.TotalRows.Value
+	if res.JobComplete != nil {
+		q.complete = res.JobComplete.Value
+	}
+	return nil
+}
+
+// Wait waits for the query to complete.
+func (q *Query) Wait(ctx context.Context) error {
+	for !q.complete {
+		err := q.checkStatus(ctx)
+		if err != nil {
+			return err
+		}
+		select {
+		case <-time.After(1 * time.Second): // TODO: exponetial backoff
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+// JobReference returns the job reference.
+func (q *Query) JobReference() *bigquerypb.JobReference {
```

Comment:
Why isn't JobReference just a member of the object?  What about stateless queries?

Reply:
good point, I can make it a member, but what happens if uses fiddle with it and set it to `nil` somehow ? And for stateless queries, the job reference is build using the `QueryID`. I need to double check if is working as intended in that part, like trying to call `AttachJob` to a stateless query.